X64 submodules zlib and googletest

chcg opened the pull request:
followup of #9
- corresponding delete\add submodules for .gitmodules adaptations
- added googletest at github as submodule, instead of deprecated gtest
- added newly zlib as submodule
- adapted vs project files therefore
- added dependency of unzip to zlib to vs solution
- corrected hardcoded link to bmp in PluginManager.rc file
- used zlib release without asm, as otherwise safeseh linker error occures with the optimized assembler versions
- maybe some vs project dependencies are still missing in vs solution
- just tested for debug and release build, to me it is unclear which of the further configuration are used and which one could be removed. I would see this as cleanup task with consistent usage of VS2013 with platform toolset v120_xp and build settings regarding compiler options/defines, usage of minimal rebuild vs. multithread build vs. precompiled headers, adding the x64 build targets.

chcg commented:
I will merge this. The issue was that I used Visual Studio "15" preview 5 aka upcoming VS2016 to build, not 2015 or 2013, which are available for production\official code. The changes between the different versions are just the platform libs to build against otherwise no build issues.

Would it be ok for you to register your github acount at appveyor, see
https://www.appveyor.com/docs/. Otherwise the CI magic will not work.

bruderstein commented:
Registered at appveyor and added the project - https://ci.appveyor.com/project/bruderstein/npppluginmanager

chcg commented:
Nice, there is still an issue with the platformset xp for zlib, I will have a look at that. V140 without xp support doesn't show this compiler warnings.

chcg commented:
Think in appveyor there are two remaining issues visible:

- Missing build dependency from Test -> googletest:

LINK : fatal error LNK1181: cannot open input file 'C:\projects\npppluginmanager\Tests\..\submodule\googletest\Win32\Release\gtest.lib' [C:\projects\npppluginmanager\Tests\Tests.vcxproj]
    15>Done Building Project "C:\projects\npppluginmanager\Tests\Tests.vcxproj" (default targets) -- FAILED.
    16>Lib:
         C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\bin\Lib.exe /OUT:"googletest\Win32\Release\gtest.lib

- Path issue with the copy task:
       PostBuildEvent:
         echo Copying C:\projects\npppluginmanager\pluginManager\projects\2010\\..\..\bin\Debug\PluginManager.dll to E:\NotepadTest\unicode\plugins
         copy "C:\projects\npppluginmanager\pluginManager\projects\2010\\..\..\bin\Debug\PluginManager.dll" "E:\NotepadTest\unicode\plugins"
         copy "C:\projects\npppluginmanager\pluginManager\bin\Debug\\*.pdb" "E:\NotepadTest\unicode\plugins"
         :VCEnd
         Copying C:\projects\npppluginmanager\pluginManager\projects\2010\\..\..\bin\Debug\PluginManager.dll to E:\NotepadTest\unicode\plugins
         The system cannot find the path specified.
                 0 file(s) copied.
         C:\projects\npppluginmanager\pluginManager\bin\Debug\\PluginManager.pdb
                 0 file(s) copied.

from paths.props:
	<TestNotepadPlusPlusUnicode>E:\NotepadTest\unicode</TestNotepadPlusPlusUnicode>
	<TestNotepadPlusPlusAnsi>e:\NotepadTest\ansi</TestNotepadPlusPlusAnsi>

Is it ok for you to modify them to the paths created in yml:

    - cd "%APPVEYOR_BUILD_FOLDER%"
    - mkdir unicodeAppDataPlugins\plugins
    - mkdir unicodeAppDataPlugins\updater

chcg commented:
Fixed the two issue.

Used just 
<TestNotepadPlusPlusUnicode>$(SolutionDir)\NotepadTest\unicode</TestNotepadPlusPlusUnicode>
and adapted yml correspondingly to check that this works for appveyor. 

@bruderstein The changed needs a confirmation\hint from your side. I don't wanna complicate your workflow, probably there is an option to modify this just for appveyor and leave it unchanged otherwise.

bruderstein commented:
The `path.props` file can't point the test N++ install to `$(SolutionDir)` - that makes the workflow really hard (after git clone, you'd need to install N++ inside the solution). But... this is meant to be a user defined path (i.e. a path to somewhere where there's an N++ that you want to test with - there will need to be a path for an x64 version too soon :) )
So, I think `path.props` should be `path.props.sample`, with a `<TestNotepadPlusPlusUnicode>c:\path\to\your\test\notepad++</TestNotepadPlusPlusUnicode>` entry, and we have a path.props.appveyor that contains the paths you need for appveyor (i.e. `$(SolutionDir)`), and we copy that file to `path.props` in the appveyor build script.

We then add `path.props` to `.gitignore`, and each dev can then setup their env without having dirty files, and maybe accidentally committing stuff. 

Also looks like appending to the $(IncludePath) can be removed from `path.props` and added to the main project, as it's always the same now, which makes the `paths.props` even simpler, and a real "user environment" file.

chcg commented:
Sounds good to me. I will do the necessary changes therefore.

Just some info to tinyxml2. I did some investigation in that direction and the most blocking point is not the slightly changed api, but that: 

All strings are query and set as 'const char*'. 

is supported for strings and here we are using TCHAR everywere aka wchar_t, if UNICODE define is active. So I think this action point has to wait a little bit in favour of x64 and x64 update capability.

chcg commented:
- removed the zlibpath and added it to the various vcxproj files, seems there should be possibility to further optimize and hide the zlib stuff behind the unzip project
- didn't remove the path.props, but added a conditional check for appveyor environment there, because otherwise loading the solution will end up with some unloaded projects. As currently there is no build description this might be a not so obvious step to other users.

bruderstein commented:
I'd really rather remove the path.props (and make it path.props.sample), and add a BUILD.md file. It's quite a common thing when there are custom things to setup.

chcg commented:
zlib\contib\minizip contains the version 1.1 of unzip, compared to current used one 1.0.1e.

with changes:

         Modifications for Zip64 support on both zip and unzip
         Copyright (C) 2009-2010 Mathias Svensson ( http://result42.com )

         For more info read MiniZip_info.txt

Original version from submodule could not be directly used, because the adaptation to TCHAR* is still needed. Probably zip64 support with more than 4GB is not relevant for the needs of the PluginManager.

bruderstein commented on appveyor.yml:
```diff
@@ -50,5 +50,6 @@ build_script:
     - cd "%APPVEYOR_BUILD_FOLDER%"
     - mkdir NotepadTest\unicode\plugins
     - mkdir NotepadTest\unicode\updater
+    - rename paths.props.appveyor paths.props
```
This should probably be `copy` - otherwise it leaves the git directory dirty. Not a big thing, but if we end up committing anything back, it will muddy the waters :)

chcg replied:
Good point, probably there will be no commit on the appveyor build, but staying in a clean mode is the better way.

bruderstein commented:
@chcg is this good to go now?

chcg commented:
From my point of view. Yes.

-- next steps would be at the configuration:
-- cleaning up the sln (I think as platform win32 should be enough,no any cpu, x86, Mixed Platforms)
-- switch all vcxproj to toolset v120_xp aka vs2013 with xp sp3 support
-- maybe rename the project dirs to 2013, what do you think
-- enable multithreaded builds, see https://randomascii.wordpress.com/2014/03/22/make-vc-compiles-fast-through-parallel-compilation/
-- maybe cleanup, optimize build settings of the vcxproj files
-- copy setting for x64 and adapt 

- for x64 plugins:
-- definition of the xml value
-- detection that x64 mode is active
-- adaptation of the parser to understand that new options
-- check/adaptation that download and installation are working
-- visual representation of the x64 results

chcg commented:
Current xml structure:

<plugin name="Plugin Manager">
	<ansiVersion>1.0.8</ansiVersion>
	<unicodeVersion>1.4.3</unicodeVersion>
	<description>The Plugin Manager is this plugin, that allows installation, update and removal of Notepad++ plugins.  	It will notify you of updates to plugins that you have installed, and automatically install dependencies and required 	files into the correct places.</description>
	<author>Dave Brotherstone</author>
	<homepage>http://www.brotherstone.co.uk/npp/pm</homepage>
	<sourceUrl>http://github.com/davegb3/nppPluginManager</sourceUrl>
	<latestUpdate>* Ignore warning about GPUP version on install *\nUpdate to use windows native internet API for downloads\nFix so that after a plugin is installed, Notepad++ will no longer be automatically promoted to run as admin - it will remain with the same privileges it had before plugin installation\nPlease report any issues to the Plugin Development forum https://notepad-plus-plus.org/community/category/5/plugin-development</latestUpdate>
	<install>
		<unicode>
			<download>https://github.com/bruderstein/nppPluginManager/releases/download/v1.4.3/PluginManager_1.4.3_UNI.zip</download>
			<copy from="plugins\PluginManager.dll" to="$PLUGINDIR$\" validate="true"/><copy from="updater\gpup.exe" to="$NPPDIR$\updater" validate="true" replace="true" isGpup="true"/>
		</unicode>
		<ansi>
		<download>http://downloads.sourceforge.net/project/npppluginmgr/v1.0.8/PluginManager_1.0.8_ANSI.zip</download>
		<copy from="plugins\PluginManager.dll" to="$PLUGINDIR$\" validate="true"/><copy from="updater\gpup.exe" to="$NPPDIR$\updater" validate="true" replace="true" isGpup="true"/>
		</ansi>
	</install>
</plugin>

- Suggestion for new X64 values:
-- <unicodeX64Version>1.4.3</unicodeX64Version>
-- <unicodeX64> download, copy </unicodeX64> within install and remove

bruderstein commented:
@chcg 
> -- switch all vcxproj to toolset v120_xp aka vs2013 with xp sp3 support

Is there still an issue with VS2015 - I've read through your notes, but I'm not quite clear on what the issue is (apologies if I'm just being slow and missing the point) - I thought with the new zlib 1.1 that would be OK?

chcg commented:
no, there is no issue with vs2015, both vs2013 and vs2015 are ok, with the highjacked and adapted version of the vcxproj.

There is a contrib dir provided together with zlib (the version we are using is 1.2.8), which contains code contributed by different developers, which is not maintained by the zlib developer himself and just there as is. Therein is a newer version 1.1 of minizip, which is included here in the unzip project in version 1.0.1e . So we have a little bit of code duplication. As the source of unzip was adapted with https://github.com/bruderstein/nppPluginManager/commit/5a5ce2929897e11f2cb8d0f9a43446ac19950a9f for tchar support, I think, we need to keep this solution or merge it with the minizip version 1.1.
